mlir/lib/Target/IRDLToCpp/Templates/PerOperationDecl.txt

```diff
@@ -0,0 +1,137 @@
+R"(
+__NAMESPACE_OPEN__
+
+namespace detail {
+
+class __OP_CPP_NAME__GenericAdaptorBase {
+public:
+  struct Properties {
+  };
+protected:
+  ::mlir::DictionaryAttr odsAttrs;
+  ::std::optional<::mlir::OperationName> odsOpName;
+  Properties properties;
+  ::mlir::RegionRange odsRegions;
+public:
+  __OP_CPP_NAME__GenericAdaptorBase(::mlir::DictionaryAttr attrs = {}, const ::mlir::EmptyProperties &properties = {}, ::mlir::RegionRange regions = {}) : odsAttrs(attrs), odsRegions(regions) {
+    if (odsAttrs)
+      odsOpName.emplace("__DIALECT_NAME__.__OP_NAME__", odsAttrs.getContext());
+  }
+
+  __OP_CPP_NAME__GenericAdaptorBase(::mlir::Operation *op) : odsAttrs(op->getRawDictionaryAttrs()), odsOpName(op->getName()), odsRegions(op->getRegions()) {}
+
+  std::pair<unsigned, unsigned> getODSOperandIndexAndLength(unsigned index, unsigned odsOperandsSize);
```

Comment:
Could you document those mysterious member functions? I think you would enter posterity if you were the one making MLIR internals readable again.